File: python/valid_sudoku.py
class Solution(object):
    def isValidSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: bool
        """
        for row in range(9):
            rows = set()
            for c in range(9):
                if board[row][c] == '.':
                    continue
                if board[row][c] in rows:
                    return False
                rows.add(board[row][c])
        
        for col in range(9):
            cols = set()
            for r in range(9):
                if board[r][col] == '.':
                    continue
                if board[r][col] in cols:
                    return False
                cols.add(board[r][col])

        for row in range(3):
            for col in range(3):
                grid = set()
                for r in range(row * 3, row * 3 + 3, 1):
                    for c in range(col * 3, col * 3 + 3, 1):
                        if board[r][c] == '.':
                            continue
                        if board[r][c] in grid:
                            return False
                        grid.add(board[r][c])
        return True

solution = Solution()
board = [
    ['5', '3', '.', '.', '7', '.', '.', '.', '.'],
    ['6', '.', '.', '1', '9', '5', '.', '.', '.'],
    ['.', '9', '8', '.', '.', '.', '.', '6', '.'],
    ['8', '.', '.', '.', '6', '.', '.', '.', '3'],
    ['4', '.', '.', '8', '.', '3', '.', '.', '1'],
    ['7', '.', '.', '.', '2', '.', '.', '.', '6'],
    ['.', '6', '.', '.', '.', '.', '2', '8', '.'],
    ['.', '.', '.', '4', '1', '9', '.', '.', '5'],
    ['.', '.', '.', '.', '8', '.', '.', '7', '9']
]

print(solution.isValidSudoku(board))

File: src/main/java/july_2025/ValidSudoku.java
package july_2025;

import java.util.HashSet;
import java.util.Set;

public class ValidSudoku {

    public static void main(String[] args) {
        char[][] board = {
                {'5', '3', '.', '.', '7', '.', '.', '.', '.'},
                {'6', '.', '.', '1', '9', '5', '.', '.', '.'},
                {'.', '9', '8', '.', '.', '.', '.', '6', '.'},
                {'8', '.', '.', '.', '6', '.', '.', '.', '3'},
                {'4', '.', '.', '8', '.', '3', '.', '.', '1'},
                {'7', '.', '.', '.', '2', '.', '.', '.', '6'},
                {'.', '6', '.', '.', '.', '.', '2', '8', '.'},
                {'.', '.', '.', '4', '1', '9', '.', '.', '5'},
                {'.', '.', '.', '.', '8', '.', '.', '7', '9'},
        };

        ValidSudoku validSudoku = new ValidSudoku();
        boolean result = validSudoku.isValidSudoku(board);
        System.out.println(result);
    }

    public boolean isValidSudoku(char[][] board) {
        for (int r = 0; r < board.length; r++) {
            Set<Integer> cols = new HashSet<>();
            for (int c = 0; c < board[r].length; c++) {
                if (board[r][c] == '.') {
                    continue;
                }
                if (cols.contains(board[r][c] - '0')) {
                    return false;
                }
                cols.add(board[r][c] - '0');
            }
        }
        for (int c = 0; c < board[0].length; c++) {
            Set<Integer> rows = new HashSet<>();
            for (int r = 0; r < board.length; r++) {
                if (board[r][c] == '.') {
                    continue;
                }
                if (rows.contains(board[r][c] - '0')) {
                    return false;
                }
                rows.add(board[r][c] - '0');
            }
        }
        for (int row = 0; row < 3; row++) {
            for (int col = 0; col < 3; col++) {
                Set<Integer> grid = new HashSet<>();
                for (int r = 3 * row; r < 3 * row + 3; r++) {
                    for (int c = 3 * col; c < 3 * col + 3; c++) {
                        if (board[r][c] == '.') {
                            continue;
                        }
                        if (grid.contains(board[r][c] - '0')) {
                            return false;
                        }
                        grid.add(board[r][c] - '0');
                    }
                }
            }
        }
        return true;
    }

}
